File: 02_algorithm/programmers/Level3/(2020-KAKAO)외벽 점검.py
import copy
import itertools

def solution(n, weak, dist):
    reverse_dist = sorted(dist, reverse=True)
    weak_length = len(weak)

    def check(perm, weak):
        criteria_weak = weak[0]
        criteria_check_weak_idx = 1
        check_available = [weak[0]]
        for dist in perm:
            for weak_idx in range(criteria_check_weak_idx, weak_length):
                if weak[weak_idx] - criteria_weak <= dist:
                    check_available.append(weak[weak_idx])
                else:
                    criteria_weak = weak[weak_idx]
                    criteria_check_weak_idx = weak_idx
                    break
        return len(check_available) == weak_length

    original_weak = copy.deepcopy(weak)
    for friend_count in range(1, len(dist) + 1):
        for dist_idx_perm in itertools.permutations(reverse_dist, friend_count):
            for _ in range(weak_length):
                if check(dist_idx_perm, weak):
                    return len(dist_idx_perm)
                pop_elem = weak.pop(0)
                weak.append(pop_elem + n)
            weak = copy.deepcopy(original_weak)
    return -1


print(solution(12, [1, 5, 6, 10], [1, 2, 3, 4])) # 2
print(solution(12, [1, 4, 6, 9], [1, 2, 3, 3])) # 2
print(solution(12, [1, 5, 6, 10], [1, 2, 3, 4])) # 2
print(solution(12, [1, 3, 4, 9, 10], [3, 5, 7])) # 1
print(solution(50, [1], [6])) # 1
print(solution(200, [0, 100], [1, 1])) # 2
print(solution(200, [0, 10, 50, 80, 120, 160], [1, 10, 5, 40, 30])) # 3
print(solution(12, [0, 10], [1, 2])) # 1
print(solution(30, [0, 3, 11, 21], [10, 4])) # 2

File: 06_js/04_algorithm/programmers/Level3/(2020-KAKAO)외벽 점검.js
function solution(n, weak, dist) {
  let answer = Infinity;
  let reverseSortedDist = dist.sort((a, b) => b - a); // (1) 내림차순으로 dist 정렬
  let weakPointCount = weak.length;

  // (5) 실제 검사하는 로직
  const isAllCheck = (perm, weakPoints) => {
    let criteriaWeak = weakPoints[0];
    let criteriaCheckWeakIdx = 1;
    let checkAvailable = [weakPoints[0]];
    for (const dist of perm) {
      for (let m = criteriaCheckWeakIdx; m < weakPointCount; m++) {
        if (weakPoints[m] - criteriaWeak <= dist) {
          checkAvailable.push(weakPoints[m]);
        } else {
          criteriaWeak = weakPoints[m];
          criteriaCheckWeakIdx = m;
          break
        }
      }
    }
    return checkAvailable.length === weakPointCount
  }

  // (4) 나온 순열을 가지고 각 취약 포인트를 시작점으로 두고
  // 모든 포인트를 검사할 수 있는지 확인하는 함수로 연결
  // ex) (n = 12) [1, 5, 8] => [5, 8, 1 + 12] => [8, 1 + 12, 5 + 12]로 변환하여
  // 계산하기 쉬운 형태로 변환
  const check = (perm) => {
    let copiedWeak = [...weak];
    for (let k = 0; k < weakPointCount; k++) {
      if (isAllCheck(perm, copiedWeak)) {
        return true;
      }
      let popElem = copiedWeak.shift();
      copiedWeak.push(popElem + n);
    }
    return false;
  }

  // (3) 순열 생성
  const makePerm = (perm, count) => {
    if (perm.length === count) {
      if (check(perm)) {
        answer = Math.min(answer, perm.length);
      }
      return
    }
    for (let j = 0; j < dist.length; j++) {
      if (!visited[j]) {
        visited[j] = true;
        perm.push(reverseSortedDist[j]);
        makePerm(perm, count);
        perm.pop();
        visited[j] = false;
      }
    }
  }

  // (2) 내림차순 정렬된 dist 원소들 중 1개 부터 dist 배열 길이 개수로
  // 뽑을 수 있는 모든 순열 구하기
  let visited = Array(dist.length).fill(false);
  for (let i = 1; i <= dist.length; i++) {
    makePerm([], i);
    if (answer !== Infinity) {
      return answer;
    }
  }
  return -1;
}

console.log(solution(12, [1, 5, 8], [1, 2, 3, 4])); // 2
console.log(solution(12, [1, 4, 6, 9], [1, 2, 3, 3])); // 2
console.log(solution(12, [1, 5, 6, 10], [1, 2, 3, 4])); // 2
console.log(solution(12, [1, 3, 4, 9, 10], [3, 5, 7])); // 1
console.log(solution(50, [1], [6])); // 1
console.log(solution(200, [0, 100], [1, 1])); // 2
console.log(solution(200, [0, 10, 50, 80, 120, 160], [1, 10, 5, 40, 30])); // 3
console.log(solution(12, [0, 10], [1, 2])); // 1
console.log(solution(30, [0, 3, 11, 21], [10, 4])); // 2